asrtoolkit/alignment/align_utils.py

```diff
@@ -0,0 +1,491 @@
+                        assert_sorted(token_matches)
+                    except AssertionError:
+                        token_matches.pop()
+        return token_matches
+
+    def unique_token_matches(self):
+        """
+        converts token_matches (List of Tuples of Ints-> aligned token ids) TO Dict
+        then convert back (List[Tuple[int,int]] --> ensures tuples are unique
+          and both hyp and ref tokens are sorted
+        """
+        token_mapper = defaultdict(dict)
+        token_matches = self.get_token_matches()
+
+        for tok1, tok2 in sorted(token_matches):
+            if tok1 not in token_mapper:
+                token_mapper[tok1] = tok2
+            else:
+                if abs(tok1 - token_mapper[tok1]) > abs(tok1 - tok2):
+                    token_mapper[tok1] = tok2
+        return token_mapper
+
+    def get_token_metadata(self):
+        """
+        given Tuples of matched token ids from hyp and ref -> self.get_token_matches()
+          merges user_data corresponding to hyp (start_time/duration/end_time)
+          and ref token of each pair (filename/speaker/gender) [requisite for STM file]
+        :return: returns list of dicts with word-level information (merged from hyp and ref)
+        """
+        token_mapper = self.unique_token_matches()
+
+        return [
+            merge(self.hyp.user_data[d1_tok], self.ref.user_data[d2_tok])
+            for d1_tok, d2_tok in token_mapper.items()
+        ]
+
+
+##########################################################################
+
+
+# Aligner Functions
+def overlap_tokens(doc, other_doc):
+    """
+    Get the tokens from the original Doc that are also in the comparison Doc.
+    """
+    overlap = []
+    other_tokens = [token.text for token in other_doc]
+    for token in doc:
+        if token.text in other_tokens:
+            other_id = other_tokens.index(token.text)
+            overlap.append([token, other_doc[other_id]])
+    return overlap
+
+
+# Sequence Match Function
+def sequence_match(gk_shared, stm_shared):  # -> List[Tuple[Span, Span]]
+    """
+    :param: gk_shared: list of Spans
+    :param: stm_shared: list of Spans
+    Spans are obtained by applying an extractor to aligned segments of text from gk_doc and stm_doc
+    and filtering results so each list only contains Spans with text present in at least one Span in the other segment
+     - thus sorted(list(set([span.text for span in {doc}_shared])))-> same for doc=gk and doc=stm
+     but the order and number of Spans in gk_shared and stm_shared may (and likely do) differ
+
+    :return: list of tuples of spans  -> List[Tuple[Span, Span]]
+        each tuple contains a matched pair from (gk_doc/segment, stm_doc/segment)
+    """
+    gk_text = [left.text for left in gk_shared]
+    stm_text = [left.text for left in stm_shared]
+
+    # Sequence Search
+    matches_list = []
+    still_left = []
+    gk_taken_list = []
+
+    for i in range(len(stm_shared)):
```

Comment:
If this loops over the two iterables zipped I don't think we'll need the `try` `except` blocks. Zip only returns pairs where both iterables exist.
```suggestion
    for stm_item, stm_word in zip(stm_shared, stm_text):
```

Reply:
100%